### ISSUE TITLE
“Process out of memory” error after installing node from source

### ISSUE THREAD

**Issue by abalter** [nodejs/help#1296]
* **Version**:
Version = v8.11.2
* **Platform**:
```
(uiserver):u89924123:~$ uname -a
Linux infongp-us78 3.16.0-ui18122.19-uiabi1-infong-amd64 #1 SMP Debian 3.16.56-1~ui80+1 (2018-05-02) x86_64 GNU/Linux
```
* **Subsystem**:

I'm trying to install NodeJS on shared hosting. I downloaded the source from 

https://nodejs.org/en/download/

and followed the usual install procedure with `./config --prefix="$HOME/local"`.

It compiles without error. However, when I try to run it, I get the following error:

    installing /kunden/homepages/32/d690510543/htdocs/local/include/node/zconf.h
    installing /kunden/homepages/32/d690510543/htdocs/local/include/node/zlib.h
    (uiserver):xxxxxxxx:~/node-v8.11.2$ which node
    /kunden/homepages/32/d690510543/htdocs/local/bin/node
    (uiserver):xxxxxxxx:~/node-v8.11.2$ node
    
    
    #
    # Fatal error in , line 0
    # API fatal error handler returned after process out of memory
    #
    Illegal instruction

However, I can get the version

    $ node -v
    v8.11.2

and help.

Nevertheless, I can't run a script

    (uiserver):xxxxxxxx:~$ cat hello.js
    var http = require('http');
    http.createServer(function (req, res) {
            res.writeHead(200, {'Content-Type': 'text/plain'});
            res.end('Hello World\n');
            }).listen(1337, "127.0.0.1");
    console.log('Server running at http://127.0.0.1:1337/');
    (uiserver):xxxxxxxx:~$ node hello.js
    
    
    #
    # Fatal error in , line 0
    # API fatal error handler returned after process out of memory
    #
    Illegal instruction

Nor can I enter the REPL:

```
(uiserver):xxxxxxxx:~$ node -i

#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction

```

**Comment by bnoordhuis** [comment]
How much memory does the machine have and how much is available to Node.js (ulimit, cgroup, etc.)?

**Comment by abalter** [comment]
```
(uiserver):xxxxxxxx:~$ free
             total       used       free     shared    buffers     cached
Mem:     132204724  128269644    3935080    1463872    1295628   98517140
-/+ buffers/cache:   28456876  103747848
Swap:       270332     103988     166344
(uiserver):xxxxxxxx:~$ ulimit
unlimited
(uiserver):xxxxxxxx:~$ vmstat
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 0  0 103988 4198496 1295636 98531376    0    0     0     0    0    0  6  1 93  0  0
(uiserver):u89924123:~$ cat /proc/meminfo
MemTotal:       132204724 kB
MemFree:         4158696 kB
MemAvailable:   123539596 kB
Buffers:         1295648 kB
Cached:         98538624 kB
SwapCached:          668 kB
Active:         61346992 kB
Inactive:       39120168 kB
Active(anon):    1001180 kB
Inactive(anon):  1098088 kB
Active(file):   60345812 kB
Inactive(file): 38022080 kB
Unevictable:        4056 kB
Mlocked:            4056 kB
SwapTotal:        270332 kB
SwapFree:         166344 kB
Dirty:               588 kB
Writeback:             0 kB
AnonPages:        636908 kB
Mapped:           231116 kB
Shmem:           1463872 kB
Slab:           23737000 kB
SReclaimable:   22347984 kB
SUnreclaim:      1389016 kB
KernelStack:       10432 kB
PageTables:        29312 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    66372692 kB
Committed_AS:    3872848 kB
VmallocTotal:   34359738367 kB
VmallocUsed:      238060 kB
VmallocChunk:   34292219416 kB
DirectMap4k:        4848 kB
DirectMap2M:     1996800 kB
DirectMap1G:    134217728 kB
```

**Comment by abalter** [comment]
Is that the buffer size? Does that seem right for shared hosting? What if I was trying to display a video or image larger than that? I'm not an expert at this.

**Comment by abalter** [comment]
And, I'm still getting the out of memory error. I'm surprised that node would be completely hobbled.

**Comment by bnoordhuis** [comment]
You could try the 32 bits binary (enjoy it while you still can, it's going away), it needs less memory on average.

That said, while Node.js has no hard lower limit, I wouldn't recommend running it with less than a few hundred MB available.

**Comment by abalter** [comment]
What do you make of this?

node --max_old_space_size=1 -i
Segmentation fault
node --max_old_space_size=2 -i
Segmentation fault
node --max_old_space_size=3 -i
terminate called after throwing an instance of 'St9bad_alloc'
  what():  std::bad_alloc
Aborted
node --max_old_space_size=4 -i
terminate called after throwing an instance of 'St9bad_alloc'
  what():  std::bad_alloc
Aborted
node --max_old_space_size=5 -i


#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction
node --max_old_space_size=6 -i


#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction
node --max_old_space_size=7 -i


#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction
node --max_old_space_size=8 -i


#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction
node --max_old_space_size=9 -i


#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction
node --max_old_space_size=10 -i


#
# Fatal error in , line 0
# API fatal error handler returned after process out of memory
#
Illegal instruction

**Comment by abalter** [comment]
If I was to ask 1&1 to make one adjustment in order to be able to run node, what would it be?

**Comment by bnoordhuis** [comment]
"Add more memory." :-)

**Comment by abalter** [comment]
TL;DR -- I'm not sure server memory limitation is actually the problem here.

It seems we found my slice has 128MB. How much do I need?  

My loop test went from `terminate called after throwing an instance of 'St9bad_alloc'` to out of memory at just 5MB.

According to [this post](https://futurestud.io/tutorials/node-js-increase-the-memory-limit-for-your-process#problem) the out of memory error should be   

    FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory  

Running in small memory situations:  
https://github.com/nodejs/help/issues/530  
https://www.fiznool.com/blog/2016/10/01/running-a-node-dot-js-app-in-a-low-memory-environment/

**Comment by abalter** [comment]
Also, `top` makes it look like I should have a decent amount of memory:

```
top - 17:24:46 up 7 days, 12:12,  2 users,  load average: 2.74, 2.84, 3.17
Tasks:   8 total,   1 running,   7 sleeping,   0 stopped,   0 zombie
%Cpu(s):  5.1 us,  0.8 sy,  0.0 ni, 94.1 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  13220472+total, 13029800+used,  1906728 free,  1275692 buffers
KiB Swap:   270332 total,   106680 used,   163652 free. 10056697+cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
17095 u899241+  22   2   22428   2864   2232 S   0.0  0.0   0:00.72 screen
17096 u899241+  22   2   16504   3476   2844 S   0.0  0.0   0:00.06 bash
20975 u899241+  22   2   84768   4220   3300 S   0.0  0.0   0:00.99 sshd
20984 u899241+  22   2   21916   5904   3200 S   0.0  0.0   0:00.03 bash
32264 u899241+  22   2   18056   2460   2092 R   0.0  0.0   0:00.00 top
33678 u899241+  22   2   84768   3956   3068 S   0.0  0.0   0:00.01 sshd
33680 u899241+  22   2   25536   9568   3256 S   0.0  0.0   0:00.05 bash
36597 u899241+  22   2   22156   2424   2192 S   0.0  0.0   0:00.00 screen
```

**Comment by abalter** [comment]
According to 1&1, my machine should have 6GB of RAM. Some something else isn't working...

**Comment by lordKnighton** [comment]
closing this issue for now, if you need anymore help let us know thanks.

**Comment by fwolfst** [comment]
For reference, I succeeded to install node on 1&1 shared hosting by installing it via nvm (like here: https://stackoverflow.com/a/33857377/1680728 ). Before, I had the same problem than the OP.
